### PR TITLE
Add precancellation chat

### PR DESCRIPTION
**Pull request by omarjackman**
# Add a precancellation chat link

This pull request adds a "chat with us" link to the purchase cancellation survey. We also add olark operator status to the redux store so that we can show and hide the chat link when operators go away/come back.

#### Test the redux store
run `npm run test-client client/state/ui/olark`

#### Cancellation survey
1. Navigate to http://calypso.localhost:3000/me/purchases/
2. Select a product to cancel
3. Click the "Remove WordPress.com Business" (or whatever the product is)
4. If operators are available you should see the chat with us link
5. Click the chat with us link
6. Notice that the olark chatbox appears in the lower right corner

### What to expect
screen shot 2016-11-09 at 9 22 18 pm

screen shot 2016-11-09 at 9 24 45 pm

### CODE REVIEW

**Comment by omarjackman** [comment]
cc @jordwest @southp @dllh

**Comment by dllh** [comment]
@evilluendas does this match visual expectations? I think your design didn't have the link underlined, and I'm not sure whether we want to avoid the underline/blue link styling or whether it's ok as is.

@omarjackman one thing we'd like to be able to do is display this for a small portion of users at a time if possible. I was planning on putting opers in a group for this purpose (but not removing them from their usual groups), and maybe enabling this chat for a small portion of users at a time, to ramp up slowly. I suspect that's going to turn out to be a difficult change to shoehorn in here. Can you think about it some and see if there's a fairly simple way? I'm wondering if we could maybe pass some context into the API call to determine eligibility, and if this is the context, we handle the campaigns differently on the server side or something.

**Comment by evilluendas** [comment]
Mmm, I would use a `.button .is-borderless` containing the whole sentence instead of a regular link, to have a bigger action area on one hand, and to keep the coherence with the other two buttons on the other. Would that work?

I think I'd also close the survey modal window when the chat is open to avoid too much stuff going on at the same time, but in that case we should probably keep the introduced values (if the user has already entered something to the survey) so they don't have to enter all the info again if they try to cancel after the chat.

File: client/state/action-types.js [hunk 1]
```diff
@@ -182,6 +182,8 @@ export const NOTICE_REMOVE = 'NOTICE_REMOVE';
 export const OLARK_READY = 'OLARK_READY';
 export const OLARK_REQUEST = 'OLARK_REQUEST';
 export const OLARK_TIMEOUT = 'OLARK_TIMEOUT';
+export const OLARK_OPERATORS_AVAILABLE = 'OLARK_OPERATORS_AVAILABLE';
+export const OLARK_OPERATORS_AWAY = 'OLARK_OPERATORS_AWAY';
```
Comment (omarjackman) [inline client/state/action-types.js]: I did and I wanted to go with the convention used for the `OLARK_READY` action. Then I figured either way works just as well as the other and I like the multi cased reducer vs the single with an `if`

```js
	switch ( action.type ) {
		case OLARK_OPERATORS_AVAILABLE:
			return OPERATOR_STATUS_AVAILABLE;
		case OLARK_OPERATORS_AWAY:
			return OPERATOR_STATUS_AWAY;
	}
	return state;
```

vs 

```js
	switch ( action.type ) {
		case OLARK_OPERATORS_STATUS_SET:
			return action.isAvailable ? OPERATOR_STATUS_AVAILABLE : OPERATOR_STATUS_AWAY;
	}
	return state;
```

**Comment by omarjackman** [comment]
@evilluendas I changed to a borderless button and also close the dialog in https://github.com/Automattic/wp-calypso/pull/9261/commits/cd0a66574dc6552b42df1d25f63111589e5a5be4

File: client/state/ui/olark/selectors.js [hunk 2]
```diff
@@ -32,3 +33,12 @@ export function isOlarkTimedOut( state ) {
 export function isRequestingOlark( state ) {
 	return state.ui.olark.requesting;
 }
+
+/**
+ * Returns if olark operators are available.
+ * @param   {Object}  state  Global state tree
+ * @returns {Boolean}        true, when olark is ready
```
Comment (omarjackman) [inline client/state/ui/olark/selectors.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/9261/commits/33185797c2bf3d1b0d5bd8d7b9773db338f98033

**Comment by omarjackman** [comment]
@dllh 

> one thing we'd like to be able to do is display this for a small portion of users at a time if possible.

The simplest way to do this is with a modulo of user id which you suggested in slack. I feel like we'll get some pushback on that but its definitely the easiest route. We could also add something to the API which would require a bit of wiring up to make it possible. What do you suggest given our time constraints?
 
> I was planning on putting opers in a group for this purpose (but not removing them from their usual groups), and maybe enabling this chat for a small portion of users at a time, to ramp up slowly.

This can be done but the solution I have in mind requires:
 * return a new value in the olark campaigns api endpoint that looks something like:
```js
groups: {
    default: "THE_DEFAULT_GROUP_ID",
    precancellation: "SPECIAL_GROUP_ID_FOR_PRECANCELATION_CHATS",
}
```
 * Store the groups in the olark redux store
 * Add actions creators that switches and keeps track of the current group before starting a chat 

This isn't very difficult but it will make this PR much larger. It might be a good candidate for v2

**Comment by dllh** [comment]
I was thinking we'd handle any modulo stuff (or whatever we land on) in the api, so that tweaking it doesn't require the PR dance. Can you characterize how much wiring it'd require? If doing it in a more maintainable, flexible way adds like a day or two, I think it's worth it because it offers us more flexibility. For example, we could in that case set up a setting that HEs can use to adjust the throttle, if desired.

Re sending multiple groups back in the endpoint, would that really scale? Like, say we later want to add a pre-sales group, or add chat in other places with other requirements and separate groups. Would we then wind up with a big list of group specifiers? I'm trying to wrap my head around why sending down a bunch of groups is a better solution than sending the server some context that tells it what group to send back for the requested context. In this case, the contact form would get one group in the response, pre-cancellation might get another (and be easy to change server-side when we're ready to open this up to all opers), and so on.

**Comment by omarjackman** [comment]
> Can you characterize how much wiring it'd require?

Lets keep this PR at its current size and I'll try to pull something together in a separate PR that handles the throttling and the requesting of a group.

> Re sending multiple groups back in the endpoint, would that really scale? 

The olark API is only hit one time at page load so my thoughts were to stick with that but we could go the route of hitting the API multiple times to get whatever group id we might be seeking. I'll spin up a separate PR for this as I mentioned above

**Comment by omarjackman** [comment]
@evilluendas @dllh @jordwest Any of you feel like theres anything left here that should stop it from shipping?

The throttling and group change will come in a later pull request

**Comment by dllh** [comment]
I gave this a test run, and it worked as expected. The only thing I'd suggest is that we hide it behind a feature flag or something until we've got the group and throttle change in, so that we don't launch before HEs are ready for it. Otherwise, I think this is :ship: ready.

**Comment by evilluendas** [comment]
Works and looks as expected for me too. Nice!

**Comment by omarjackman** [comment]
> The only thing I'd suggest is that we hide it behind a feature flag

Its hidden here: https://github.com/Automattic/wp-calypso/pull/9261/files#diff-28dcdb0cfe9db785ba5043118ab70f42R194

**Pull request merged**